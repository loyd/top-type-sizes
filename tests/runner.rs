use structopt::StructOpt;

use top_type_sizes::*;

fn test(content: &'static str) {
    let refined_content = reader::read(content.as_bytes()).unwrap();

    let types = parser::parse(&refined_content).unwrap();
    insta::with_settings!({ description => "internal representation" }, {
        insta::assert_yaml_snapshot!("types", types);
    });

    snap_output(&types, &[]);
    snap_output(&types, &["-ws"]);
    snap_output(&types, &["-h8"]);
    snap_output(&types, &["-ws", "-h16"]);
}

fn snap_output(types: &[schema::Type], cmd: &[&str]) {
    let full_cmd = std::iter::once(&"top-type-sizes")
        .chain(cmd)
        .copied()
        .collect::<Vec<_>>();

    let options = options::Options::from_iter(&full_cmd);
    let snap_name = {
        let mut parts = cmd.to_vec();
        parts.insert(0, "output");
        parts.join("")
    };

    let types = transformer::transform(types.to_vec(), &options);
    let output = formatter::format(types, &options);

    insta::with_settings!({ description => full_cmd.join(" ") }, {
        insta::assert_snapshot!(snap_name, output);
    })
}

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            insta::with_settings!(
            {
                snapshot_path => concat!("snapshots/", stringify!($name)),
                prepend_module_to_snapshot => false,
                omit_expression => true,
            },
            {
                test(include_str!(concat!("samples/", stringify!($name), ".txt")))
            })
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
