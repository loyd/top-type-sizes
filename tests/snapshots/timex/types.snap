---
source: tests/runner.rs
description: internal representation
---
- name: "unix::linux_like::linux::gnu::timex"
  size: 208
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: modes
            size: 4
        - Padding: 4
        - Field:
            kind: AdtField
            name: offset
            size: 8
            align: 8
        - Field:
            kind: AdtField
            name: freq
            size: 8
        - Field:
            kind: AdtField
            name: maxerror
            size: 8
        - Field:
            kind: AdtField
            name: esterror
            size: 8
        - Field:
            kind: AdtField
            name: status
            size: 4
        - Padding: 4
        - Field:
            kind: AdtField
            name: constant
            size: 8
            align: 8
        - Field:
            kind: AdtField
            name: precision
            size: 8
        - Field:
            kind: AdtField
            name: tolerance
            size: 8
        - Field:
            kind: AdtField
            name: time
            size: 16
        - Field:
            kind: AdtField
            name: tick
            size: 8
        - Field:
            kind: AdtField
            name: ppsfreq
            size: 8
        - Field:
            kind: AdtField
            name: jitter
            size: 8
        - Field:
            kind: AdtField
            name: shift
            size: 4
        - Padding: 4
        - Field:
            kind: AdtField
            name: stabil
            size: 8
            align: 8
        - Field:
            kind: AdtField
            name: jitcnt
            size: 8
        - Field:
            kind: AdtField
            name: calcnt
            size: 8
        - Field:
            kind: AdtField
            name: errcnt
            size: 8
        - Field:
            kind: AdtField
            name: stbcnt
            size: 8
        - Field:
            kind: AdtField
            name: tai
            size: 4
        - Field:
            kind: AdtField
            name: __unused1
            size: 4
        - Field:
            kind: AdtField
            name: __unused2
            size: 4
        - Field:
            kind: AdtField
            name: __unused3
            size: 4
        - Field:
            kind: AdtField
            name: __unused4
            size: 4
        - Field:
            kind: AdtField
            name: __unused5
            size: 4
        - Field:
            kind: AdtField
            name: __unused6
            size: 4
        - Field:
            kind: AdtField
            name: __unused7
            size: 4
        - Field:
            kind: AdtField
            name: __unused8
            size: 4
        - Field:
            kind: AdtField
            name: __unused9
            size: 4
        - Field:
            kind: AdtField
            name: __unused10
            size: 4
        - Field:
            kind: AdtField
            name: __unused11
            size: 4
